homepage/models.py

```diff
@@ -99,12 +99,71 @@ def get_details(self):
 
+    # --- returns if course has Prerequisites
+    def has_preqs(self):
+        return Prerequisites.does_course_have_prerequisites(self)
+
     # --- returns all Course objects - the main 'courses' source for the view
     @staticmethod
     def get_courses():
         return Course.objects.all()
 
-    # --- get filtered course list by rating/load/mandatory/elective  - return QuerySets
+    # --- filters 'curr_courses' and returns the result according to 'filters'
+    @staticmethod
+    def get_filtered_courses(curr_courses, filters):
+        active_filters = []
+        for filter in filters:
+            if(filter == 'mand'):
+                curr_courses = Course.get_mandatory_courses(curr_courses)
+                active_filters.append('mandatory')
+            elif(filter == 'elect'):
+                curr_courses = Course.get_elective_courses(curr_courses)
+                active_filters.append('elective')
+            elif(filter == 'load_below'):
+                curr_courses = Course.get_filtered_courses_by_load(3.5, curr_courses)
+                active_filters.append('course load under 3.5')
+            elif(filter == 'rate_over'):
+                curr_courses = Course.get_filtered_courses_by_rating(3.5, curr_courses)
+                active_filters.append('course rating over 3.5')
+            elif(filter == 'has_preqs'):
+                curr_courses = Course.get_courses_with_preqs(curr_courses)
+                active_filters.append('with prerequisites')
+            elif(filter == 'no_preqs'):
+                curr_courses = Course.get_courses_without_preqs(curr_courses)
+                active_filters.append('without prerequisites')
+            elif(filter == 'rater_num'):
+                curr_courses = Course.get_courses_with_ratings(curr_courses, 5)
+                active_filters.append('at least 5 raters')
+
+        return {'result': curr_courses, 'active': active_filters}
+
+    # --- sorts 'curr_courses' and return the result according to 'sort_val'
+    @staticmethod
+    def get_sorted_courses(curr_courses, sort_val):
+        current_sorting = ''
+        if(sort_val == 'name'):
+            curr_courses = Course.sort_by_name(curr_courses)
+            current_sorting = 'name'
+        elif(sort_val == 'id'):
+            curr_courses = Course.sort_by_id(curr_courses)
+            current_sorting = 'identifier'
+        elif(sort_val == 'rating'):
+            curr_courses = Course.sort_by_rating(curr_courses)
+            current_sorting = 'course rating'
+        elif(sort_val == 'load'):
+            curr_courses = Course.sort_by_load(curr_courses)
+            current_sorting = 'course load'
+        elif(sort_val == 'num_reviews'):
+            curr_courses = Course.sort_by_num_reviews(curr_courses)
+            current_sorting = 'number of reviews'
+        elif(sort_val == 'num_raters'):
+            curr_courses = Course.sort_by_num_raters(curr_courses)

```

Comment:
maybe you can use "case-switch" instead of 'if-else" conditions IMHO it looks more readable 
```suggestion
match sort_val:
    case "name":
            curr_courses = Course.sort_by_name(curr_courses)
            current_sorting = 'name'
    case "id":
            curr_courses = Course.sort_by_id(curr_courses)
            current_sorting = 'identifier'
...
and so on
```

Reply:
From what I can tell it would require updating to python 3.10, which just came out lately and isn't active currently, according to Python.org